projects/RabbitMQ.Client/client/impl/AsyncRpcContinuations.cs

```diff
@@ -31,26 +31,70 @@
 
 using System;
 using System.Runtime.CompilerServices;
+using System.Threading;
 using System.Threading.Tasks;
 using RabbitMQ.Client.client.framing;
 using RabbitMQ.Client.Exceptions;
 using RabbitMQ.Client.Framing.Impl;
 
 namespace RabbitMQ.Client.Impl
 {
-    internal abstract class AsyncRpcContinuation<T> : IRpcContinuation
+    internal abstract class AsyncRpcContinuation<T> : IRpcContinuation, IDisposable
     {
+        private readonly CancellationTokenSource _ct;
+
         protected readonly TaskCompletionSource<T> _tcs = new TaskCompletionSource<T>(TaskCreationOptions.RunContinuationsAsynchronously);
 
+        private bool _disposedValue;
+
+        public AsyncRpcContinuation(TimeSpan continuationTimeout)
+        {
+            _ct = new CancellationTokenSource(continuationTimeout);
+            _ct.Token.Register(() =>
+            {
+                if (_tcs.TrySetCanceled())
+                {
+                    // TODO LRB #1347
+                    // Cancellation was successful, does this mean we should set a TimeoutException
+                    // in the same manner as BlockingCell?
```

Comment:
Look at [this link](https://learn.microsoft.com/en-us/dotnet/standard/parallel-programming/task-cancellation) for guidance on Task Cancellation patterns. Technically, I think all of the async methods should accept a CancellationToken (with a default value of None) so people can easily set their own cancellation timeouts. It's also good practice to wrap those with a TimeSpan overload that just calls the method with a Cancellationtoken forwarding the TimeSpan parameter.

Reply:
That makes sense, thanks for taking a look.

Reply:
Of course, you can't cancel an operation once sent to RabbitMQ, so really what we need to support is a continuation timeout in the same way that `BlockingCell` does. I'll see what other .NET libraries do with regard to async operations and timeouts.